Feat/o2k improvements

dimitropoulos opened the pull request:
A recent PR (https://github.com/Kong/insomnia/pull/3340) on inso went through a lot of iterations to try and find a suitable solution to the problem, and during that time there were some small things fixed along the way.  I pulled those things out of the work on that PR and collected them here in the interest of cleanliness (but also because it looks like that PR may not merge super soon).

Best to read commit by commit, I tried to keep it very clean and document any changes where necessary.

dimitropoulos committed:
adds overloads so that types of package exports

it would be FAR preferable to use an object with the arguments instead because then we don't have to fall into all the traps that come with overloads, but I am avoiding changing the actual call signature of these function.

dimitropoulos committed:
fixes and updates types

dimitropoulos committed:
corrects mispelling

dimitropoulos committed:
removes shim types and exports actual types from openapi-2-kong

we had everything in place for the rest of the build pipeline to consume these types, but were missing the main exports at the root index.ts file

dimitropoulos committed:
use actual openapi-2-kong types now that they're available

dimitropoulos committed:
makes this `yaml` import match all others

for better or for worse, until we upgrade `YAML` (it's now written in TypeScript) we should probably aim for consistency.

dimitropoulos committed:
use existing variable for filePath

dimitropoulos committed:
use KubernetesManifest for name of kubernetes manifest union

when first naming this, I didn't have much more context than the property name (documents), so after taking a look,  it's clear that these are kubernetes manifests

dimitropoulos committed:
removes attempt at correctly typing overloads

I will return to this, and so I'm leaving it here in a commit rather than squashing it out, but I can't see a way to get it to work without changing the signature in a big way.

dimitropoulos committed:
fixes error (now that there are yaml types)

develohpanda commented on packages/openapi-2-kong/src/declarative-config/upstreams.test.ts:
```diff
@@ -4,6 +4,7 @@ import { getSpec } from './utils';
 
 const xKongUpstreamDefaults = 'x-kong-upstream-defaults';
 
+/** This function is written in such a way as to allow mutations in tests but without affecting other tests. */
```
👍🏽

develohpanda commented on packages/openapi-2-kong/src/types/outputs.ts:
```diff
@@ -1,32 +1,26 @@
 
 export type ConversionResultType = 'kong-declarative-config' | 'kong-for-kubernetes';
 
+export interface Warnings {
```
nit pick, this should be a singular `Warning` right?

dimitropoulos replied:
most definitely.  great catch.  Will update.

dimitropoulos replied:
updated in https://github.com/Kong/insomnia/pull/3396/commits/33643d2d4f81079b1fe5e199a965740f62341e36

develohpanda commented on packages/insomnia-inso/src/commands/generate-config.test.ts:
```diff
@@ -14,7 +14,8 @@ const generate = _generate as jest.MockedFunction<typeof _generate>;
 
-const mockConversionResult: ConversionResult = {
+const mockConversionResult: KongForKubernetesResult = {
+  // @ts-expect-error -- TSCONVERSION the tests seem to suggest that this is valid, yet it is not allowed by the types.
```
Yeah I believe this is because the document is just a string rather than a complete K8s document. This was done for simplicity (because flow wasn't very helpful) but we can look to enforce it better now.

dimitropoulos replied:
that's good to know, I don't want to get into it quite yet because I'd like to have a little more time to look at/understand how it works for these functions taking strings that represent json or yaml for valid specs, or objects that represent valid specs themselves.

ultimately, a package where we can have common types, basically we can use https://github.com/Kong/khcp/blob/master/apps/konnect-api/src/common/globals.ts#L23 as a jumping off point.  That way, we'll be able to (at least a little better) describe the contents of these strings.  it's a little bigger than what I want to dip into for this PR, but just to share what's on my mind.

develohpanda commented on packages/openapi-2-kong/src/types/kubernetes-config.ts:
```diff
@@ -79,6 +79,6 @@ export interface KubernetesConfig {
 }
 
-export type KubernetesDocument = KubernetesConfig
+export type KubernetesManifest = KubernetesConfig
```
👍🏽

develohpanda reviewed (commented):
Need to QA so not approving as yet but code review checks out

dimitropoulos committed:
updates mistake in name per review feedback

dimitropoulos committed:
makes `K8s` vs `Kubernetes` type terminology consistent

part of the motivation for this is that there are significant things in common between the two kinds of configs, making them closer in form is therefore ideal.

dimitropoulos committed:
extracts o2k-specific helper data to live in o2k

evidently, this data is useful for the usage of o2k, therefore o2k should be the thing exporting it.

develohpanda reviewed (approved):
![](https://media3.giphy.com/media/5t1YusAYxmRn474Sid/giphy.gif)
